### PR TITLE
Introduce MLv2 rename-join function

### PR DESCRIPTION
**Pull request by metamben**
Part of #30093.

### CODE REVIEW

File: frontend/src/metabase-lib/join.ts [hunk 1]
```diff
@@ -98,3 +98,12 @@ export function joinFields(join: Join): ColumnMetadata[] {
 export function withJoinFields(join: Join, newFields: ColumnMetadata[]): Join {
   return ML.with_join_fields(join, newFields);
 }
+
+export function renameJoin(
+  query: Query,
+  stageIndex: number,
+  oldNameOrIndex: string | number,
```
Comment (metamben) [inline frontend/src/metabase-lib/join.ts:105]: The target can be specified by its alias or index. Other option would be requiring the whole join clause. Input welcome!

File: modules/drivers/athena/test/metabase/driver/athena_test.clj [hunk 2]
```diff
@@ -81,8 +81,12 @@
...
                              (assoc-in [:middleware :format-rows?] false))]
         (mt/with-native-query-testing-context query
-          (is (= [#t "2022-11-16T04:21:00.000-08:00[America/Los_Angeles]" #t "05:03"]
-                 (mt/first-row (qp/process-query query)))))))))
+          (let [[ts t] (mt/first-row (qp/process-query query))]
+            (is (#{#t "2022-11-16T04:21:00.000-08:00[America/Los_Angeles]"
+                   #t "2022-11-16T04:21:00.000-08:00[US/Pacific]"}
+                   ts))
+            (is (= #t "05:03"
+                   t))))))))
```
Comment (snoe) [inline modules/drivers/athena/test/metabase/driver/athena_test.clj:89]: belongs elsewhere?

File: src/metabase/lib/join.cljc [hunk 3]
```diff
@@ -584,3 +584,49 @@
...
+                              idx))
+                          (m/indexed the-joins)))
+        old-name  (get-in the-joins [idx :alias])]
+    (if (and idx (not= old-name new-name))
+      (let [unique-name-fn (lib.util/unique-name-generator)
+            _              (run! unique-name-fn (map :alias the-joins))
+            unique-name    (unique-name-fn new-name)]
+        (-> stage
+            (assoc-in [:joins idx :alias] unique-name)
+            (update-in [:joins idx] replace-join-alias old-name unique-name)))
```
Comment (snoe) [inline src/metabase/lib/join.cljc:613]: I thought join field references can exist with a `:join-alias` elsewhere within the stage?

e.g.
```clojure
{:joins [{... :alias "foo"}]
 :filters [[:= {...} [:field {:join-alias "foo"} 123] 1]]}
 ```

**Review by snoe — APPROVED** [review]

File: modules/drivers/athena/test/metabase/driver/athena_test.clj (hunk 2 above)
Comment (metamben) [reply modules/drivers/athena/test/metabase/driver/athena_test.clj:89]: Yes, there was a failing Athena test, the fix has been merged in a separate PR.

**Pull request merged**